eth2/utils/merkle_proof/src/lib.rs

```diff
@@ -73,6 +85,68 @@ impl MerkleTree {
         }
     }
 
+    /// Push an element in the MerkleTree.
+    /// MerkleTree and depth must be correct, as the algorithm expects valid data.
+    pub fn push_leaf(&mut self, elem: H256, depth: usize) -> Result<(), MerkleTreeError> {
+        use MerkleTree::*;
+
+        if depth == 0 {
+            return Err(MerkleTreeError::DepthTooSmall);
+        }
+
+        let mut right;
+        let mut left;
+        match &*self {
+            Leaf(_) => return Err(MerkleTreeError::LeafReached),
+            Zero(_) => *self = MerkleTree::create(&[elem], depth),
+            Node(_, l, r) => {
+                match (&**l, &**r) {
+                    // Tree is full
+                    (Leaf(_), Leaf(_)) => return Err(MerkleTreeError::MerkleTreeFull),
+                    // There is a right node so insert in right node
+                    (Node(_, _, _), Node(_, _, _)) => {
+                        right = *r.clone();
+                        if let Err(e) = right.push_leaf(elem, depth - 1) {
+                            return Err(e);
+                        }
+                        left = *l.clone();
+                    }
+                    // Both branches are zero, insert in left one
+                    (Zero(_), Zero(_)) => {
+                        left = MerkleTree::create(&[elem], depth - 1);
+                        right = *r.clone();
+                    }
+                    // Leaf on left branch and zero on right branch, insert on right side
+                    (Leaf(_), Zero(_)) => {
+                        left = *l.clone();
+                        right = MerkleTree::create(&[elem], depth - 1);
+                    }
+                    // Try inserting on the left node -> if it fails because it is full, insert in right side.
```

Comment:
This is a pretty funky binary tree insertion algorithm, but like you said, it's still `O(log n)`, so we should be alright.

Reply:
It is indeed. As I said, we get a big improvement in terms of inserting in the case of `(Node(_), Zero(_))`, so I believe it's worth it!